### PR TITLE
fix: web navigation broken, when going back and forth in history

### PR DESCRIPTION
**Pull request by HeAtNet**
This fix resolves the #9128 issue
https://github.com/react-navigation/react-navigation/issues/9128

Until not merged, you can try this, by adding the following code to the package.json file's scripts property, and it will use the fixed code for @react-navigation/native.
`"postinstall": "git clone --branch web-navigation-fix https://github.com/HeAtNet/react-navigation.git temp_react_navigation && cd temp_react_navigation/packages/native && npm install @types/jest --save-dev && npm i && npm run prepare && rm node_modules -r -f && cd ../../.. && rm node_modules/@react-navigation/native -r -f && cp temp_react_navigation/packages/native node_modules/@react-navigation/ -r && rm temp_react_navigation -r -f",`

### CODE REVIEW

**Comment by muqker** [comment]
This did indeed solve the issue for me. I hope the pull request gets accepted. For now, I am using it as a patch-package.

**Comment by TikiTDO** [comment]
@satya164 Any chance this PR could get a review? This fixes a really nasty web navigation issue. The only task failing for this PR is the `welcome` github action, which appears to be erroring out immediately.

**Comment by satya164** [comment]
I'll have some time to test this next week.

**Comment by adamhari** [comment]
Hoping this makes it in soon, I am anxious to support browser navigation with react-native-web. This library works really well apart from that!

File: packages/native/src/useLinking.tsx [hunk 1]
```diff
@@ -46,9 +46,16 @@ const createMemoryHistory = () => {
...
+        let currentIndex = items.findIndex((item) => item.id === id);
+        if (currentIndex < 0) {
+          currentIndex = 0;
+        }
 
-        return index > -1 ? index : 0;
+        // Fix createMemoryHistory.index variable's value, because it will go out of sync, when navigating
+        // with the browser. The 'popstate' event not sets the createMemoryHistory.index variable, but
+        // calls this function, so the corrrect value can be set.
+        index = currentIndex;
```
Comment (satya164) [inline packages/native/src/useLinking.tsx:57]: We shouldn't change the internal state inside a getter. Please move this logic to the `onPopState` function which is the appropriate place.

File: packages/native/src/useLinking.tsx [hunk 2]
```diff
@@ -95,7 +102,12 @@ const createMemoryHistory = () => {
 
       const id = window.history.state?.id ?? nanoid();
 
-      if (items.length) {
+      if (items.findIndex((item) => item.id === id) < 0) {
```
Comment (satya164) [inline packages/native/src/useLinking.tsx:105]: This is also going to be `true` when `items.length` is`0`, causing the `else` clause in this condition to never get called. Should this check be after the `items.length` check?

**Review by satya164 — CHANGES_REQUESTED** [review]
Thanks for the PR. Can you address the review? I'll merge after they are addressed.

**Comment by Hirbod** [comment]
@HeAtNet could you add the requested change requests by satya?

File: packages/native/src/useLinking.tsx (hunk 1 above)
Comment (HeAtNet) [reply packages/native/src/useLinking.tsx]: Yes, that is the appropriate place. I fixed.

File: packages/native/src/useLinking.tsx (hunk 2 above)
Comment (HeAtNet) [reply packages/native/src/useLinking.tsx]: That's true. Furthermore the third statement is not needed, so I changed the conditions.

**Comment by HeAtNet** [comment]
I made the requested changes.
The integration test failed and I don't know why, but I hope this is a good solution now.

**Comment by TikiTDO** [comment]
@HeAtNet You may want to merge in main (and perhaps target main as the merge target).

**Comment by takacskalman** [comment]
Any updates?

**Comment by PofMagicfingers** [comment]
I think @HeAtNet needs to change the merge target to main, for this to work according to @TikiTDO

**Comment by Hirbod** [comment]
@satya164 can you work with this PR or would you like me to create a new PR targeting main? This one is sitting here for a while now and web navigation with browser navigation buttons are broken ever since.

**Comment by satya164** [comment]
Thank you for updating the PR. I rebased it and merged it in fb84805

**Comment by Hirbod** [comment]
While this fix works fine while navigation back and forth now, it still throws a crash when you reach the initial screen. 
It works perfectly fine when you click back and forth on any other screen.
Bildschirmfoto 2021-09-28 um 11 20 42

**Comment by PofMagicfingers** [comment]
I think we need to open a new issue, this one is not tracked anymore

**Pull request closed without merging**